### PR TITLE
[ios, macos] Add Mapbox Maps SDK metrics manager.

### PR DESCRIPTION
**Pull request by fabian-guerra**
@julianrex this is a new concept I'm exploring. The difference from our last conversation is that I'm using blocks to pass the event using a block (similar to our lightweight logging handler approach) rather than implementing a protocol.

This sends the event to our default telemetry implementation and the block handler.

```
- (void)handleEvent:(NSString *)eventName withAttributes:(NSDictionary *)attributes {
#if TARGET_OS_IPHONE || TARGET_OS_SIMULATOR
    [MGLMapboxEvents pushEvent:eventName withAttributes:attributes];
#endif
    if (self.shouldHandleEvents && self.handler) {
        self.handler(eventName, attributes);
    }
}
```
This does not duplicates the `MGLMapboxEvents` class and keeps there our telemetry calls, while sharing the event with our developers. If someone does not wants to share this event with us they just have to turn off the telemetry library, and still they will get the event. Which keeps consistency with other events.

### CODE REVIEW

**Comment by fabian-guerra** [comment]
The last commits (https://github.com/mapbox/mapbox-gl-native/pull/13997/commits/610a07505d80c4492ded6b436685f083847fd9e2, https://github.com/mapbox/mapbox-gl-native/pull/13997/commits/1ae1a0c1fcf86bca0be4ebde0b33a8e54cf891e2) explore a different approach where taking inspiration from [NSURLSessionTaskMetrics](https://developer.apple.com/documentation/foundation/nsurlsessiontaskmetrics?language=objc) we define `MGLMetricsManager` that will let sdk clients push the `metrics event` to our backend using the Telemetry library, and `MGLMetricsDelegate` let developers handle `metrics events` generated in the sdk. For now we have the performance event only.

File: platform/darwin/src/MGLMetricsManager.m [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#import "MGLMetricsManager_Private.h"
+#import "MGLNetworkConfiguration_Private.h"
+#if TARGET_OS_IOS
+#import "MGLMapboxEvents.h"
+#endif
+
+NSString* MGLStringFromMetricType(MGLMetricType metricType) {
+    return [@[kMGLDownloadPerformanceEvent] objectAtIndex:metricType];
```
Comment (julianrex) [inline platform/darwin/src/MGLMetricsManager.m]: In this case we could just return `kMGLDownloadPerformanceEvent` and check that `metricType == 0`. We could return `nil` if not.

Ideally, we should check to ensure that the the size of the array and the count of `MGLMetricType`s are the same.

File: platform/darwin/src/MGLMetricsManager.m [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+@implementation MGLMetricsManager
+
++ (instancetype)sharedManager
+{
+    static dispatch_once_t once;
+    static id sharedConfiguration;
+    dispatch_once(&once, ^{
+        sharedConfiguration = [[self alloc] init];
+        [MGLNetworkConfiguration sharedManager].metricsDelegate = sharedConfiguration;
```
Comment (julianrex) [inline platform/darwin/src/MGLMetricsManager.m]: Ideally we'd try to remove this explicit dependency - but let's think about that as follow on work.

File: platform/darwin/src/MGLMetricsManager.m [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+#if TARGET_OS_IOS
+- (void)pushMetric:(MGLMetricType)metricType withAttributes:(NSDictionary *)attributes {
+    
+    NSString *eventName;
+    
+    switch (metricType) {
+        case MGLMetricTypePerformance:
+            eventName = kMGLDownloadPerformanceEvent;
```
Comment (julianrex) [inline platform/darwin/src/MGLMetricsManager.m]: Can you use `MGLStringFromMetricType` for this?

File: platform/darwin/src/MGLNetworkConfiguration.m [hunk 4]
```diff
@@ -14,6 +20,7 @@ + (instancetype)sharedManager {
     dispatch_once(&onceToken, ^{
         _sharedManager = [[self alloc] init];
         _sharedManager.sessionConfiguration = nil;
+        _sharedManager.events = [NSMutableDictionary dictionary];
```
Comment (julianrex) [inline platform/darwin/src/MGLNetworkConfiguration.m]: This should really be part of `MGLNetworkConfiguration`'s initializer - let's move into an `init`

File: platform/darwin/src/MGLNetworkConfiguration.m [hunk 5]
```diff
@@ -48,4 +55,53 @@ - (NSURLSessionConfiguration *)defaultSessionConfiguration {
     return sessionConfiguration;
 }
 
+- (void)startDownloadEvent:(NSString *)urlString type:(NSString *)resourceType {
+    if (urlString && ![self.events objectForKey:urlString]) {
+        [self.events setObject:@{ MGLStartTime: [NSDate date], MGLResourceType: resourceType } forKey:urlString];
```
Comment (julianrex) [inline platform/darwin/src/MGLNetworkConfiguration.m]: Is it possible to have multiple events with the same `urlString`?

File: platform/darwin/src/MGLMetricsManager.m [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            break;
+    }
+    
+    [MGLMapboxEvents pushEvent:eventName withAttributes:attributes];
+
+}
+#endif
+
+- (void)networkConfiguration:(MGLNetworkConfiguration *)networkConfiguration didGenerateMetricEvent:(NSDictionary *)metricEvent {
+    [self handleMetricsEvent:MGLMetricTypePerformance withAttributes:metricEvent];
```
Comment (julianrex) [inline platform/darwin/src/MGLMetricsManager.m]: We should pull the metric type from the `event` value in `metricEvent` here.

File: platform/darwin/src/MGLNetworkConfiguration.m (hunk 5 above)
Comment (julianrex) [inline platform/darwin/src/MGLNetworkConfiguration.m]: Should we use the upcoming `MMEDate` here @rclee?

Comment (rclee) [reply platform/darwin/src/MGLNetworkConfiguration.m]: Going to refer to @alfwatt on this one but I don't think it should make much difference.

Comment (fabian-guerra) [reply platform/darwin/src/MGLNetworkConfiguration.m]: At different times yes.

File: platform/darwin/src/MGLMetricsManager.m (hunk 6 above)
Comment (fabian-guerra) [reply platform/darwin/src/MGLMetricsManager.m]: Why?

File: platform/darwin/src/MGLMetricsManager.m (hunk 2 above)
Comment (fabian-guerra) [reply platform/darwin/src/MGLMetricsManager.m]: Sure. Just a remainder that `MGLNetworkConfiguration` is part of another target and there should not be explicit dependencies.

File: platform/darwin/src/MGLNetworkConfiguration.m (hunk 5 above)
Comment (julianrex) [reply platform/darwin/src/MGLNetworkConfiguration.m]: But not overlapping?

File: platform/darwin/src/MGLMetricsManager.m (hunk 6 above)
Comment (julianrex) [reply platform/darwin/src/MGLMetricsManager.m]: Can/could `networkConfiguration: didGenerateMetricEvent:` be called with another event type?

File: platform/darwin/src/MGLNetworkConfiguration.m (hunk 5 above)
Comment (fabian-guerra) [reply platform/darwin/src/MGLNetworkConfiguration.m]: No at the same time. You may start downloading a tile and if is not cached then you will download that tile again but at a different time.

File: platform/darwin/src/MGLMetricsManager.m (hunk 6 above)
Comment (fabian-guerra) [reply platform/darwin/src/MGLMetricsManager.m]: No.

File: platform/darwin/src/http_file_source.mm [hunk 7]
```diff
@@ -215,14 +215,19 @@ void cancel() {
...
+        
         request->task = [impl->session
             dataTaskWithRequest:req
               completionHandler:^(NSData* data, NSURLResponse* res, NSError* error) {
                 if (error && [error code] == NSURLErrorCancelled) {
+                    [[MGLNetworkConfiguration sharedManager] cancelDownloadEvent:res.URL.relativePath];
                     return;
                 }
-
+                [[MGLNetworkConfiguration sharedManager] stopDownloadEvent:res.URL.relativePath];
```
Comment (julianrex) [inline platform/darwin/src/http_file_source.mm]: We're not currently distinguishing between success and error here. What does Android do here?

File: platform/darwin/src/MGLNetworkConfiguration.m [hunk 8]
```diff
@@ -48,4 +62,53 @@ - (NSURLSessionConfiguration *)defaultSessionConfiguration {
...
+    }
+}
+
+- (NSDictionary *)eventAttributesForURL:(NSString *)urlString withAction:(NSString *)action
+{
+    NSDictionary *parameters = [self.events objectForKey:urlString];
+    NSDate *startDate = [parameters objectForKey:MGLStartTime];
+    NSDate *endDate = [NSDate date];
+    NSTimeInterval elapsedTime = [endDate timeIntervalSinceDate:startDate];
+    NSDateFormatter* iso8601Formatter = [[NSDateFormatter alloc] init];
```
Comment (julianrex) [inline platform/darwin/src/MGLNetworkConfiguration.m]: For later: we should look at supporting `NSISO8601DateFormatter` where available.

File: platform/darwin/src/MGLMetricsManager.m [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    
+    [MGLMapboxEvents pushEvent:eventName withAttributes:mutableAttributes];
+}
+
+NSString* deviceName()
+{
+    struct utsname systemInfo;
+    uname(&systemInfo);
+    
+    return [NSString stringWithCString:systemInfo.machine
```
Comment (julianrex) [inline platform/darwin/src/MGLMetricsManager.m]: Note for future: `systemInfo.nodename` is also interesting - on my iPhoneX this is `iPhone-X`.

**Pull request merged**